### PR TITLE
Added transforms for up- and downsampling

### PR DESCRIPTION
**Pull request by mknitt**
Adds transforms and tests for up- and downsampling of time series data.

Required for #15

### CODE REVIEW

File: src/flowcean/transforms/downsample.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    "group_by_dynamic". The method can downsample to the first, last or mean
+    value of the group.
+
+    Args:
+        time_column (str): The name of the time column.
+
+        sampling_rate (str | timedelta): The sampling rate to downsample to.
+
+        value (str): The value to use for the downsampling. Can be "first",
+            "last" or "mean".
```
Comment (schmidma) [inline src/flowcean/transforms/downsample.py:24]:
```suggestion
        time_column: The name of the time column.
        sampling_rate: The sampling rate to downsample to.
        value: The value to use for the downsampling. Can be "first", "last" or "mean".
```

File: src/flowcean/transforms/downsample.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+class Downsample(Transform):
+    """Downsample a time series.
+
+    Downsample a time series to a lower sampling rate using the polars method
+    "group_by_dynamic". The method can downsample to the first, last or mean
+    value of the group.
+
+    Args:
+        time_column (str): The name of the time column.
```
Comment (schmidma) [inline src/flowcean/transforms/downsample.py:19]: we usually call these "columns" _features_

File: src/flowcean/transforms/downsample.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        self.time_column = time_column
+        self.value = value
+
+    def transform(self, data: pl.DataFrame) -> pl.DataFrame:
+        logger.debug("Downsampling time series")
+
+        if self.value == "first":
+            data = data.group_by_dynamic(
+                self.time_column, every=self.sampling_rate
+            ).agg(
```
Comment (schmidma) [inline src/flowcean/transforms/downsample.py:45]: You may be able to do this `group_by_dynamic` call once, as this is lazy in polars and then subsequently do the different aggregations afterwards.

Or maybe even better, first select the proper query for each version like
```
aggretation = pl.col(x)..... for x in ...
```
and then always do the
```
data = data.groupby().agg(aggregation)
```

File: src/flowcean/transforms/downsample.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+                pl.col(x).last() for x in data.columns if x != self.time_column
+            )
+        elif self.value == "mean":
+            data = data.group_by_dynamic(
+                self.time_column, every=self.sampling_rate
+            ).agg(
+                pl.col(x).mean() for x in data.columns if x != self.time_column
```
Comment (schmidma) [inline src/flowcean/transforms/downsample.py:60]: I think there is something in polars to select all columns except one. Maybe this https://docs.pola.rs/py-polars/html/reference/expressions/api/polars.exclude.html

File: src/flowcean/transforms/upsample.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    def transform(self, data: pl.DataFrame) -> pl.DataFrame:
+        logger.debug("Upsampling time series")
+        return (
+            data.upsample(
+                time_column=self.time_column,
+                every=self.sampling_rate,
+                offset=self.offset,
+            )
+            .interpolate()
+            .fill_null(strategy="forward")
```
Comment (schmidma) [inline src/flowcean/transforms/upsample.py:43]: expose the strategy to the user by providing it as an argument?

File: src/flowcean/transforms/upsample.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    """Upsample a time series.
+
+    Upsample a time series to a higher sampling rate using the polars method
+    "upsample". The method can upsample to a higher sampling rate and fill
+    missing values using interpolation.
+
+    Args:
+        time_column (str): The name of the time column.
+        sampling_rate (str | timedelta): The sampling rate to upsample to.
+        offset (str | timedelta): The offset to use for the upsampling.
```
Comment (schmidma) [inline src/flowcean/transforms/upsample.py:21]:
```suggestion
        time_column: The name of the time column.
        sampling_rate: The sampling rate to upsample to.
        offset: The offset to use for the upsampling.
```

File: src/flowcean/transforms/upsample.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class Upsample(Transform):
+    """Upsample a time series.
+
+    Upsample a time series to a higher sampling rate using the polars method
+    "upsample". The method can upsample to a higher sampling rate and fill
+    missing values using interpolation.
```
Comment (schmidma) [inline src/flowcean/transforms/upsample.py:16]:
```suggestion
    missing values using an interpolation strategy.
```

**Review by schmidma — CHANGES_REQUESTED** [review]
some minor remarks and suggestions

**Comment by mknitt** [comment]
Closing this PR because the functionality is already provided by #122

**Pull request closed without merging**